src/base/LruMap.h

```diff
@@ -148,50 +184,70 @@ LruMap<EntryValue, EntryCost>::add(const char *key, EntryValue *t)
     if (memLimit() == 0)
         return false;
 
-    index.push_front(new Entry(key, t));
-    storage.insert(MapPair(key, index.begin()));
+    // optimization: save on memory usage for already-stale objects
+    // (findEntry() would discard instead later anyway)
+    if (ttl < 0)
+        return false;
+
+    auto e = new Entry(key, t, ttl);
+    memUsed_ += e->memoryUsed();
+    index.push_front(e);
+
+    auto m = MapPair(key, index.begin());
+    memUsed_ += memoryUsedBy(m);
+    storage.insert(m);
 
     ++entries_;
-    return true;
+
+    // if over memory limit, start purging
+    trim();
```

Comment:
> Unless the entry by itself exceeds the total map storage space it cannot be added then immediately purged.

Correct. It is that "unless" case that (as you can probably tell from my sketch) I am worried about, and that the reviewed code did not handle well AFAICT.